storage_splunk_hec.go

```diff
@@ -16,34 +18,56 @@ import (
 // storage backend.
 type SplunkHecConfig struct {
 	Token                     string `yaml:"token"`
-	Tenant                    string `yaml:"tenant"`
-	Port                      int    `yaml:"port"`
+	HecURL                    string `yaml:"hec-url"`
 	Host                      string `yaml:"host"`
 	Source                    string `yaml:"source"`
 	MetricsSourceType         string `yaml:"metrics-source-type"`
 	MetricsIndex              string `yaml:"metrics-index"`
 	EventsSourceType          string `yaml:"events-source-type"`
 	EventsIndex               string `yaml:"events-index"`
 	SkipCertificateValidation bool   `yaml:"skip-cert-validation"`
+	CaCert                    string `yaml:"ca-cert"`
 }
 
 // SplunkHecStorage holds the configuration of a Splunk HEC storage backend
 type SplunkHecStorage struct {
 	client *http.Client
 	config SplunkHecConfig
-	url    string
 	ctx    context.Context
 }
 
 // NewSplunkHecStorage sets up a new Splunk HEC storage backend
 func NewSplunkHecStorage(c *Config) (SplunkHecStorage, error) {
+	var s SplunkHecStorage
+	s.config = c.Storage.SplunkHec
 	tr := &http.Transport{
 		Proxy:                 http.ProxyFromEnvironment,
 		TLSHandshakeTimeout:   10 * time.Second,
 		ExpectContinueTimeout: 1 * time.Second,
 	}
-	if c.Storage.SplunkHec.SkipCertificateValidation {
-		tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	if c.Storage.SplunkHec.CaCert != "" {
+		rootCAs, err := x509.SystemCertPool()
+		if err != nil {
+			return s, fmt.Errorf("unable to load system certificate pool: %v", err)
+		}
+		if rootCAs == nil {
+			return s, fmt.Errorf("unable to append ca-cert, system certificate pool is nil")
+		}
+		certs, err := ioutil.ReadFile(c.Storage.SplunkHec.CaCert)
+		if err != nil {
+			return s, fmt.Errorf("unable to load ca-cert from %s: %v", c.Storage.SplunkHec.CaCert, err)
+		}
+		if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
+			log.Printf("unable to append ca-cert from %s, using system certs only\n", c.Storage.SplunkHec.CaCert)
+		}
+		tr.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: c.Storage.SplunkHec.SkipCertificateValidation,
+			RootCAs:            rootCAs,
+		}
+	} else {
+		tr.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: c.Storage.SplunkHec.SkipCertificateValidation,
```

Comment:
if no cert was provided, we know we're going to have to skip cert verification (because we know they self-sign).  We shouldn't make the user have to configure this.  Just skip and log a message saying we skipped.